extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DefaultTokenIntrospectionUserInfoCache.java

```diff
@@ -0,0 +1,186 @@
+package io.quarkus.oidc.runtime;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import io.quarkus.oidc.OidcTenantConfig;
+import io.quarkus.oidc.TokenIntrospection;
+import io.quarkus.oidc.TokenIntrospectionCache;
+import io.quarkus.oidc.UserInfo;
+import io.quarkus.oidc.UserInfoCache;
+import io.smallrye.mutiny.Uni;
+
+/**
+ * Default TokenIntrospection and UserInfo Cache implementation.
+ * A single cache entry can keep TokenIntrospection and/or UserInfo.
+ * 
+ * In most cases it is the opaque bearer access tokens which are introspected
+ * but the code flow access tokens can also be introspected if they have the roles claims.
+ *
+ * In either case, if a remote request to fetch UserInfo is required then it will be the same access token
+ * which has been introspected which will be used to request UserInfo.
+ */
+@ApplicationScoped
+public class DefaultTokenIntrospectionUserInfoCache implements TokenIntrospectionCache, UserInfoCache {
```

Comment:
@stuartwdouglas I have a test where a custom cache is preferred - it uses an Alternative priority and it is doc-ed so I think it should be fine - the same is the case with `TokenStateManager` and users have tried registering custom ones.

> Also because it implements both TokenIntrospectionCache and UserInfoCache even if this was fixed this would still cause issues if the user only wanted to replace one.

This is a bigger problem. Not sure what to do as I wanted to avoid keeping 2 separate storages (and timers) for the default cache, I'll need to think about it, may be it is simpler to have a single interface and provide a no-op abstract support

Reply:
I've realized it should be fine with `@AltPriority` as well. I've thought about it and I'm not sure it would be great to have an interface like `TokenIntrospectionUserInfoCache`. `@AltPriority` is a simple way to customize - `DefaultTokenIntrospectionUserInfoCache` is registered but is not activated OOB, and if possible I'd like to avoid adding another `enabled` property - as it already works the same way with the default token state manager and the users are successfully `replacing` it with `@AltPriority`.
And if for example one wants to write a custom factory for `TokenIntrospection` only but reuse the default UserInfo cache implementation (unlikely IMHO, the custom cache would use Redis, Infinispan and aligning it with the basic memory based cache does not make much sense) then one can also implement both interfaces but delegate to the injected default factory as needed